File: AllJoyn/Samples/OICAdapter/iotivity-1.0.0/android/examples/devicediscoveryclient/src/main/java/org/iotivity/base/examples/DeviceDiscoveryClient.java
/*
 * //******************************************************************
 * //
 * // Copyright 2015 Intel Corporation.
 * //
 * //-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
 * //
 * // Licensed under the Apache License, Version 2.0 (the "License");
 * // you may not use this file except in compliance with the License.
 * // You may obtain a copy of the License at
 * //
 * //      http://www.apache.org/licenses/LICENSE-2.0
 * //
 * // Unless required by applicable law or agreed to in writing, software
 * // distributed under the License is distributed on an "AS IS" BASIS,
 * // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * // See the License for the specific language governing permissions and
 * // limitations under the License.
 * //
 * //-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
 */
package org.iotivity.base.examples;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.text.method.ScrollingMovementMethod;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.ScrollView;
import android.widget.TextView;

import org.iotivity.base.ModeType;
import org.iotivity.base.OcConnectivityType;
import org.iotivity.base.OcException;
import org.iotivity.base.OcPlatform;
import org.iotivity.base.OcRepresentation;
import org.iotivity.base.PlatformConfig;
import org.iotivity.base.QualityOfService;
import org.iotivity.base.ServiceType;

import java.util.EnumSet;
import java.util.HashMap;
import java.util.Map;

/**
 * This sample demonstrates the device discovery feature.
 * The client queries for the device related information stored by the server.
 */
public class DeviceDiscoveryClient extends Activity implements
        OcPlatform.OnDeviceFoundListener,
        OcPlatform.OnPlatformFoundListener {
    private void startDeviceDiscoveryClient() {
        Context context = this;

        PlatformConfig platformConfig = new PlatformConfig(
                context,
                ServiceType.IN_PROC,
                ModeType.CLIENT,
                "0.0.0.0", // By setting to "0.0.0.0", it binds to all available interfaces
                0,         // Uses randomly available port
                QualityOfService.LOW
        );

        msg("Configuring platform.");
        OcPlatform.Configure(platformConfig);
        sleep(1);

        try {
            msg("Querying for platform information...");
            OcPlatform.getPlatformInfo("",
                    OcPlatform.WELL_KNOWN_PLATFORM_QUERY,
                    EnumSet.of(OcConnectivityType.CT_DEFAULT),
                    this);
        } catch (OcException e) {
            Log.e(TAG, e.toString());
            msg("Failed to query for platform information");
        }
        sleep(2);

        try {
            msg("Querying for device information...");
            OcPlatform.getDeviceInfo("",
                    OcPlatform.WELL_KNOWN_DEVICE_QUERY,
                    EnumSet.of(OcConnectivityType.CT_DEFAULT),
                    this);
        } catch (OcException e) {
            Log.e(TAG, e.toString());
            msg("Failed to query for device information");
        }
        sleep(2);

        enableStartButton();
        printLine();
    }

    private final static Map<String, String> PLATFORM_INFO_KEYS = new HashMap<String, String>() {{
        put("pi", "Platform ID: ");
        put("mnmn", "Manufacturer name: ");
        put("mnml", "Manufacturer url: ");
        put("mnmo", "Manufacturer Model No: ");
        put("mndt", "Manufactured Date: ");
        put("mnpv", "Manufacturer Platform Version: ");
        put("mnos", "Manufacturer OS version: ");
        put("mnhw", "Manufacturer hardware version: ");
        put("mnfv", "Manufacturer firmware version: ");
        put("mnsl", "Manufacturer support url: ");
        put("st", "Manufacturer system time: ");
    }};

    @Override
    public synchronized void onPlatformFound(OcRepresentation ocRepresentation) {
        msg("Platform Information received:");
        try {
            for (String key : PLATFORM_INFO_KEYS.keySet()) {
                msg("\t" + PLATFORM_INFO_KEYS.get(key) + ocRepresentation.getValue(key));
            }
        } catch (OcException e) {
            Log.e(TAG, e.toString());
            msg("Failed to read platform info values.");
        }

        printLine();
    }

    private final static Map<String, String> DEVICE_INFO_KEYS = new HashMap<String, String>() {{
        put("di", "Device ID: ");
        put("n", "Device name: ");
        put("lcv", "Spec version url: ");
        put("dmv", "Data Model: ");
    }};

    @Override
    public synchronized void onDeviceFound(OcRepresentation ocRepresentation) {
        msg("Device Information received:");
        try {
            for (String key : DEVICE_INFO_KEYS.keySet()) {
                msg("\t" + DEVICE_INFO_KEYS.get(key) + ocRepresentation.getValue(key));
            }
        } catch (OcException e) {
            Log.e(TAG, e.toString());
            msg("Failed to read device info values.");
        }

        printLine();
    }

    //******************************************************************************
    // End of the OIC specific code
    //******************************************************************************

    private final static String TAG = DeviceDiscoveryClient.class.getSimpleName();
    private TextView mConsoleTextView;
    private ScrollView mScrollView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_device_discovery_client);

        mConsoleTextView = (TextView) findViewById(R.id.consoleTextView);
        mConsoleTextView.setMovementMethod(new ScrollingMovementMethod());
        mScrollView = (ScrollView) findViewById(R.id.scrollView);
        mScrollView.fullScroll(View.FOCUS_DOWN);
        final Button button = (Button) findViewById(R.id.button);

        if (null == savedInstanceState) {
            button.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    button.setText("Re-start");
                    button.setEnabled(false);
                    new Thread(new Runnable() {
                        public void run() {
                            startDeviceDiscoveryClient();
                        }
                    }).start();
                }
            });
        } else {
            String consoleOutput = savedInstanceState.getString("consoleOutputString");
            mConsoleTextView.setText(consoleOutput);
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString("consoleOutputString", mConsoleTextView.getText().toString());
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);

        String consoleOutput = savedInstanceState.getString("consoleOutputString");
        mConsoleTextView.setText(consoleOutput);
    }

    private void enableStartButton() {
        runOnUiThread(new Runnable() {
            public void run() {
                Button button = (Button) findViewById(R.id.button);
                button.setEnabled(true);
            }
        });
    }

    private void sleep(int seconds) {
        try {
            Thread.sleep(seconds * 1000);
        } catch (InterruptedException e) {
            e.printStackTrace();
            Log.e(TAG, e.toString());
        }
    }

    private void msg(final String text) {
        runOnUiThread(new Runnable() {
            public void run() {
                mConsoleTextView.append("\n");
                mConsoleTextView.append(text);
                mScrollView.fullScroll(View.FOCUS_DOWN);
            }
        });
        Log.i(TAG, text);
    }

    private void printLine() {
        msg("------------------------------------------------------------------------");
    }
}

File: AllJoyn/Samples/OICAdapter/iotivity-1.0.0/resource/examples/devicediscoveryclient.cpp
//******************************************************************
//
// Copyright 2014 Intel Mobile Communications GmbH All Rights Reserved.
//
//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

///
///This sample demonstrates the device discovery feature
///The client queries for the device related information
///stored by the server.
///

#include <mutex>
#include <condition_variable>

#include "OCPlatform.h"
#include "OCApi.h"

using namespace OC;

static void printUsage()
{
    std::cout << "Usage devicediscoveryclient <0|1>" << std::endl;
    std::cout << "connectivityType: Default IP" << std::endl;
    std::cout << "connectivityType 0: IP" << std::endl;
}
//Callback after device information is received
void receivedPlatformInfo(const OCRepresentation& rep)
{
    std::cout << "\nPlatform Information received ---->\n";
    std::string value;
    std::string values[] =
    {
        "pi",   "Platform ID                    ",
        "mnmn", "Manufacturer name              ",
        "mnml", "Manufacturer url               ",
        "mnmo", "Manufacturer Model No          ",
        "mndt", "Manufactured Date              ",
        "mnpv", "Manufacturer Platform Version  ",
        "mnos", "Manufacturer OS version        ",
        "mnhw", "Manufacturer hardware version  ",
        "mnfv", "Manufacturer firmware version  ",
        "mnsl", "Manufacturer support url       ",
        "st",   "Manufacturer system time       "
    };

    for (unsigned int i = 0; i < sizeof(values) / sizeof(values[0]) ; i += 2)
    {
        if(rep.getValue(values[i], value))
        {
            std::cout << values[i + 1] << " : "<< value << std::endl;
        }
    }
}

void receivedDeviceInfo(const OCRepresentation& rep)
{
    std::cout << "\nDevice Information received ---->\n";
    std::string value;
    std::string values[] =
    {
        "di",   "Device ID        ",
        "n",    "Device name      ",
        "lcv",  "Spec version url ",
        "dmv",  "Data Model Model ",
    };

    for (unsigned int i = 0; i < sizeof(values) / sizeof(values[0]) ; i += 2)
    {
        if(rep.getValue(values[i], value))
        {
            std::cout << values[i + 1] << " : "<< value << std::endl;
        }
    }
}

int main(int argc, char* argv[]) {

    std::ostringstream platformDiscoveryRequest;
    std::ostringstream deviceDiscoveryRequest;

    std::string platformDiscoveryURI = "/oic/p";
    std::string deviceDiscoveryURI   = "/oic/d";

    //Default Connectivity type
    OCConnectivityType connectivityType = CT_ADAPTER_IP;

    if(argc == 2)
    {
        try
        {
            std::size_t inputValLen;
            int optionSelected = std::stoi(argv[1], &inputValLen);

            if(inputValLen == strlen(argv[1]))
            {
                if(optionSelected == 0)
                {
                    std::cout << "Using IP."<< std::endl;
                    connectivityType = CT_ADAPTER_IP;
                }
                else
                {
                    std::cout << "Invalid connectivity type selected." << std::endl;
                    printUsage();
                    return -1;
                }
            }
            else
            {
                std::cout << "Invalid connectivity type selected. Using default IP" << std::endl;
            }
        }
        catch(std::exception&)
        {
            std::cout << "Invalid input argument. Using IP as connectivity type" << std::endl;
        }
    }
    else
    {
        printUsage();
    }
    // Create PlatformConfig object
    PlatformConfig cfg {
        OC::ServiceType::InProc,
        OC::ModeType::Client,
        "0.0.0.0",
        0,
        OC::QualityOfService::LowQos
    };

    OCPlatform::Configure(cfg);
    try
    {
        platformDiscoveryRequest << OC_MULTICAST_PREFIX << platformDiscoveryURI;
        deviceDiscoveryRequest << OC_MULTICAST_PREFIX << deviceDiscoveryURI;

        OCStackResult ret;

        std::cout<< "Querying for platform information... ";

        ret = OCPlatform::getPlatformInfo("", platformDiscoveryRequest.str(), connectivityType,
                &receivedPlatformInfo);

        if (ret == OC_STACK_OK)
        {
            std::cout << "done." << std::endl;
        }
        else
        {
            std::cout << "failed." << std::endl;
        }

        std::cout<< "Querying for device information... ";

        ret = OCPlatform::getDeviceInfo("", deviceDiscoveryRequest.str(), connectivityType,
                &receivedDeviceInfo);

        if (ret == OC_STACK_OK)
        {
            std::cout << "done." << std::endl;
        }
        else
        {
            std::cout << "failed." << std::endl;
        }

        // A condition variable will free the mutex it is given, then do a non-
        // intensive block until 'notify' is called on it.  In this case, since we
        // don't ever call cv.notify, this should be a non-processor intensive version
        // of while(true);
        std::mutex blocker;
        std::condition_variable cv;
        std::unique_lock<std::mutex> lock(blocker);
        cv.wait(lock, []{return false;});

    }catch(OCException& e)
    {
        std::cerr << "Failure in main thread: "<<e.reason()<<std::endl;
    }

    return 0;
}


